# Text Classification Project
Now we're at the point where we should be able to:
* Read in a collection of documents - a *corpus*
* Transform text into numerical vector data using a pipeline
* Create a classifier
* Fit/train the classifier
* Test the classifier on new data
* Evaluate performance

For this project we'll use the Cornell University Movie Review polarity dataset v2.0 obtained from http://www.cs.cornell.edu/people/pabo/movie-review-data/

In this notebook we'll try to develop a classification model as we did for the SMSSpamCollection dataset - that is, we'll try to predict the Positive/Negative labels based on text content alone.

## Perform imports and load the dataset
The dataset contains the text of 2000 movie reviews. 1000 are positive, 1000 are negative, and the text has been preprocessed as a tab-delimited file.

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('../TextFiles/moviereviews.tsv', sep = '\t')

In [2]:
df.head()

label                                             review
0   neg  how do films like mouse hunt get into theatres...
1   neg  some talented actresses are blessed with a dem...
2   pos  this has been an extraordinary year for austra...
3   pos  according to hollywood movies made in last few...
4   neg  my first press screening of 1998 and already i...

In [3]:
len(df)

2000

### Take a look at a typical review. This one is labeled "negative":

In [4]:
df.iloc[0,1]

'how do films like mouse hunt get into theatres ? \r\nisn\'t there a law or something ? \r\nthis diabolical load of claptrap from steven speilberg\'s dreamworks studio is hollywood family fare at its deadly worst . \r\nmouse hunt takes the bare threads of a plot and tries to prop it up with overacting and flat-out stupid slapstick that makes comedies like jingle all the way look decent by comparison . \r\nwriter adam rifkin and director gore verbinski are the names chiefly responsible for this swill . \r\nthe plot , for what its worth , concerns two brothers ( nathan lane and an appalling lee evens ) who inherit a poorly run string factory and a seemingly worthless house from their eccentric father . \r\ndeciding to check out the long-abandoned house , they soon learn that it\'s worth a fortune and set about selling it in auction to the highest bidder . \r\nbut battling them at every turn is a very smart mouse , happy with his run-down little abode and wanting it to stay that way . \r\

* NaN records are efficiently handled with [.isnull()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.isnull.html) and [.dropna()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.dropna.html)
* Strings that contain only whitespace can be handled with [.isspace()](https://docs.python.org/3/library/stdtypes.html#str.isspace), [.itertuples()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.itertuples.html), and [.drop()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

Lets see the `NaN` values. We can delete these by using `dropna()`

In [5]:
df[df['review'].isnull()]

label review
140    pos    NaN
208    pos    NaN
270    neg    NaN
334    neg    NaN
448    neg    NaN
522    neg    NaN
606    pos    NaN
696    neg    NaN
728    pos    NaN
738    neg    NaN
744    pos    NaN
786    pos    NaN
820    pos    NaN
866    neg    NaN
986    neg    NaN
1072   neg    NaN
1156   pos    NaN
1204   pos    NaN
1260   neg    NaN
1272   neg    NaN
1294   neg    NaN
1314   pos    NaN
1330   neg    NaN
1332   neg    NaN
1426   pos    NaN
1456   pos    NaN
1486   pos    NaN
1538   neg    NaN
1568   neg    NaN
1584   pos    NaN
1588   pos    NaN
1600   pos    NaN
1662   pos    NaN
1904   pos    NaN
1908   neg    NaN

In [6]:
df.isnull().sum()

label      0
review    35
dtype: int64

In [7]:
df.dropna(inplace=True)

In [8]:
len(df)

1965

We correctly dropped the 35 missing values.

We can use the `isspace()` to extract out the reviews with empty space.

In [9]:
blanks = []

for i, label, review in df.itertuples():
    if type(review) == str:
        if review.isspace():
            blanks.append(i)
print(len(blanks), 'blanks', blanks)

27 blanks [57, 71, 147, 151, 283, 307, 313, 323, 343, 351, 427, 501, 633, 675, 815, 851, 977, 1079, 1299, 1455, 1493, 1525, 1531, 1763, 1851, 1905, 1993]


Next we'll pass our list of index numbers to the **.drop()** method, and set `inplace=True` to make the change permanent.

In [10]:
df.drop(blanks,inplace=True)

In [11]:
len(df)

1938

Great, we dropped 62 blank records from the original 2000.

## Take a quick look at the `label` column:

In [12]:
df['label'].value_counts()

neg    969
pos    969
Name: label, dtype: int64

## Split the data into train & test sets:

In [13]:
from sklearn.model_selection import train_test_split

In [14]:
X = df['review']
y = df['label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

## Build pipelines to vectorize the data, then train and fit a model
Now that we have sets to train and test, we'll develop a selection of pipelines, each with a different model.

We will build two models 
- Naive Bayes
- Linear Support Vector Classifier

In [15]:
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import LinearSVC

# Naive Bayes
text_clf_nb = Pipeline([('tfidf', TfidfVectorizer()),
                       ('clf', MultinomialNB())])

# Linear SVC

text_clf_lsvc = Pipeline([('tfidf', TfidfVectorizer()),
                          ('clf', LinearSVC())])



## Feed the training data through the first pipeline (naïve Bayes)

In [16]:
text_clf_nb.fit(X_train,y_train)

Pipeline(memory=None,
     steps=[('tfidf', TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.float64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), norm='l2', preprocessor=None, smooth_idf=True,...rue,
        vocabulary=None)), ('clf', MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))])

## Run predictions and analyze the results (naïve Bayes)

In [17]:
prediction_nb = text_clf_nb.predict(X_test)

In [18]:
from sklearn import metrics

print(metrics.confusion_matrix(y_test, prediction_nb))

[[287  21]
 [130 202]]


In [19]:
# Print a classification report
print(metrics.classification_report(y_test,prediction_nb))

              precision    recall  f1-score   support

         neg       0.69      0.93      0.79       308
         pos       0.91      0.61      0.73       332

   micro avg       0.76      0.76      0.76       640
   macro avg       0.80      0.77      0.76       640
weighted avg       0.80      0.76      0.76       640



In [20]:
# Print the overall accuracy
print(metrics.accuracy_score(y_test,prediction_nb))

0.7640625


Naïve Bayes gave us better-than-average results at 76.4% for classifying reviews as positive or negative based on text alone. Let's see if we can do better.

## Feed the training data through the second pipeline (Linear SVC)

In [21]:
text_clf_lsvc.fit(X_train, y_train)

Pipeline(memory=None,
     steps=[('tfidf', TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.float64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), norm='l2', preprocessor=None, smooth_idf=True,...ax_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0))])

In [22]:
prediction_svc = text_clf_lsvc.predict(X_test)

In [23]:
from sklearn import metrics

print(metrics.confusion_matrix(y_test, prediction_svc))

[[259  49]
 [ 49 283]]


In [24]:
print(metrics.classification_report(y_test, prediction_svc))

              precision    recall  f1-score   support

         neg       0.84      0.84      0.84       308
         pos       0.85      0.85      0.85       332

   micro avg       0.85      0.85      0.85       640
   macro avg       0.85      0.85      0.85       640
weighted avg       0.85      0.85      0.85       640



In [25]:
print(metrics.accuracy_score(y_test, prediction_svc))

0.846875


Based on text alone we correctly classified reviews as positive or negative **84.7%** of the time. In an upcoming section we'll try to improve this score even further by performing *sentiment analysis* on the reviews.